README.md
<!DOCTYPE html>
<html>
<head>
</head>
<body>

![auto-ts](logo.png)

<h1 id="auto-ts">Auto_TS: Auto_TimeSeries</h1>
<p style="font-family:verdana">Automatically build multiple Time Series models using a Single Line of Code.</p>
<p>Auto_timeseries is a complex model building utility for time series data. Since it automates many
Tasks involved in a complex endeavor, it assumes many intelligent defaults. But you can change them.
Auto_Timeseries will rapidly build predictive models based on Statsmodels ARIMA, Seasonal ARIMA
and Scikit-Learn ML. It will automatically select the best model which gives best score specified.
</p>
<p>New version 0.0.25 onwards changes the syntax of Auto_TimeSeries to be more like scikit-learn (fit and predict syntax). You will have to initialize an object and then call fit with your data and then predict again with data. Hope this makes it easier to remember and use.</p>
<h2 id="introduction">Introduction</h2>
<p>Auto_TimeSeries enables you to build and select multiple time series models using techniques such as ARIMA, SARIMAX, VAR, decomposable (trend+seasonality+holidays) models, and ensemble machine learning models.</p>
<p>Auto_TimeSeries is an Automated ML library for time series data. Auto_TimeSeries was initially conceived and developed by <a href="https://www.linkedin.com/in/ram-seshadri-nyc-nj/">Ram Seshadri</a> and was significantly expanded in functionality and scope and upgraded to its present status by <a href="https://github.com/ngupta23">Nikhil Gupta</a>.</p>
<p>auto-ts.Auto_TimeSeries() is the main function that you will call with your train data. You can then choose what kind of models you want: stats, ml or FB prophet based model. You can also tell it to automatically select the best model based on the scoring parameter you want it to be based on. It will return the best model and a dictionary containing predictions for the number of forecast_periods you mentioned (default=2).</p>
<h2 id="installation-instructions">INSTALLATION INSTRUCTIONS</h2>
<ol>
<li>Use “pip install auto-ts”</li>
<li>Use “pip3 install auto-ts” if the above doesn’t work</li>
<li>pip install git+git://github.com/AutoViML/Auto_TS</li>
</ol>
<h2 id="run">RUN</h2>
<p> First you need to import auto_timeseries from auto_ts library:<br>
<code>
from auto_ts import auto_timeseries
</code>
</p>
<p>
Second, Initialize an auto_timeseries model object which will hold all your parameters:</p>
<p><code>
model = auto_timeseries(
            score_type='rmse',
            time_interval='Month',
            non_seasonal_pdq=None, seasonality=False,
            seasonal_period=12,
            model_type=['Prophet'],
            verbose=2)
</code></p>
Here are how the input parameters defined:<br>
<ol>
<li><b>score_type (default='rmse')</b>: The metric used for scoring the models. Type is string.
Currently only the following two types are supported:
(1) "rmse": Root Mean Squared Error (RMSE)
(2) "normalized_rmse": Ratio of RMSE to the standard deviation of actuals</li>
<li><b>time_interval (default is None)</b>: Used to indicate the frequency at which the data is collected.
This is used for two purposes (1) in building the Prophet model and (2) used to impute the seasonal period for SARIMAX in case it is not provided by the user (None). Type is String.
We use the following <a href="https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases">pandas date range frequency</a> aliases that Prophet uses to make the prediction dataframe. <p>Hence, please note that these are the list of allowed aliases for frequency:
                      ['B','C','D','W','M','SM','BM','CBM',
                     'MS','SMS','BMS','CBMS','Q','BQ','QS','BQS',
                     'A,Y','BA,BY','AS,YS','BAS,BYS','BH',
                     'H','T,min','S','L,ms','U,us','N']

For a start, you can test the following codes for your data and see how the results are:
<ul>
<li>'MS', 'M', 'SM', 'BM', 'CBM', 'SMS', 'BMS' for monthly frequency data</li>
<li> 'D', 'B', 'C' for daily frequency data </li>
<li> 'W' for weekly frequency data </li>
<li> 'Q', 'BQ', 'QS', 'BQS' for quarterly frequency data </li>
<li> 'A,Y', 'BA,BY', 'AS,YS', 'BAS,YAS' for yearly frequency data </li>
<li> 'BH', 'H', 'h' for hourly frequency data </li>
<li> 'T,min' for minute frequency data </li>
<li> 'S', 'L,milliseconds', 'U,microseconds', 'N,nanoseconds' for second frequency data </li>
</ul>
Or you can leave it as None and auto_timeseries will try and impute it.
</li>
<li><b>non_seasonal_pdq (default = (3,1,3))</b>: Indicates the maximum value of (p, d, q) to be used in the search for statistical ARIMA models.
If None, then the following values are assumed max_p = 3, max_d = 1, max_q = 3. Type is Tuple.</li>
<li><b>seasonality (default=False)</b>: Used in the building of the SARIMAX model only at this time. True or False. Type is bool.</li>
<li><b>seasonal_period (default is None)</b>: Indicates the seasonal period in your data. This depends on the peak (or valley) period that occurs regularly in your data.
Used in the building of the SARIMAX model only at this time.
There is no impact of this argument if seasonality is set to False
If None, the program will try to infer this from the time_interval (frequency) of the data
We assume the following as defaults but feel free to change them.
(1) If frequency is Monthly, then seasonal_period is assumed to be 12
(1) If frequency is Daily, then seasonal_period is assumed to be 30 (but it could be 7)
(1) If frequency is Weekly, then seasonal_period is assumed to be 52
(1) If frequency is Quarterly, then seasonal_period is assumed to be 4
(1) If frequency is Yearly, then seasonal_period is assumed to be 1
(1) If frequency is Hourly, then seasonal_period is assumed to be 24
(1) If frequency is Minutes, then seasonal_period is assumed to be 60
(1) If frequency is Seconds, then seasonal_period is assumed to be 60
Type is integer</li>
<li><b>conf_int (default=0.95)</b>: Confidence Interval for building the Prophet model. Default: 0.95. Type is float.</li>
<li><b>model_type (default: 'stats'</b>: The type(s) of model to build. Default to building only statistical models
Can be a string or a list of models. Allowed values are:
'best', 'prophet', 'stats', 'ARIMA', 'SARIMAX', 'VAR', 'ML'.
"prophet" will build a model using FB Prophet -> this means you must have FB Prophet installed
"stats" will build statsmodels based ARIMA, SARIMAX and VAR models
"ML" will build a machine learning model using Random Forests provided explanatory vars are given
'best' will try to build all models and pick the best one
If a list is provided, then only those models will be built
WARNING: "best" might take some time for large data sets. We recommend that you
choose a small sample from your data set before attempting to run entire data.
Type can be either: [string, list]
</li>
<li><b>verbose (default=0)</b>: Indicates the verbosity of printing (Default = 0). Type is integer.</li>
</ol>
The next step after defining the model object is to fit it with some real data:
<p>
<code>
model.fit(
            traindata=train_data,
            ts_column=ts_column,
            target=target,
            cv=5,
            sep=","
        )
</code></p>
<br>Here are how the parameters defined:
<ul>
<li><b>traindata (required)</b>: It can be either a dataframe or a file. You must give the name of the file along with its data path in case if a file. It also accepts a pandas dataframe in case you already have a dataframe loaded in your notebook.</li>
<li><b>ts_column (required)</b>: name of the datetime column in your dataset (it could be a name of column or index number in the columns index)</li>
<li><b>target (required)</b>: name of the column you   are trying to predict. Target could also be the only column in your dataset</li>
<li><b>cv (default=5)</b>: You can enter any integer for the number of folds you want in your cross validation data set.
</li>
<li><b>sep (default=",")</b>: Sep is the separator in your traindata file. If your separator is ",", "\t", ";", make sure you enter it here. If not, it is ignored.</li>
</ul>
The next step after training the model object is to make some predictions with test data:<br>
<p>
<code>
predictions = model.predict(
            testdata = can be either a dataframe or an integer signifying the forecast_period,
            model = 'best' or any other string that stands for the trained model
        )  
</code></p>
Here are how the parameters are defined. You can choose to send either testdata in the form of a dataframe or send in an integer to decide how many periods you want to forecast.  You need only
<ul>
<li><b>testdata (required)</b>: It can be either a dataframe containing test data or you can use an integer standing for the forecast_period (you want).</li>
<li><b>model (optional, default = 'best')</b>: The name of the model you want to use among the many different models you have trained. Remember that the default is the best model. But you can choose any model that you want to forecast with. Type is String.</li>
</ul>
<h2 id="requirements">Requirements</h2>
<p>tscv</p>
<p>scikit-learn</p>
<p>FB Prophet</p>
<p>statsmodels</p>
<p>pmdarima</p>
<h1 id="license">License:</h1>
<p>Apache License 2.0</p>
<h2 id="recommendations">Recommendations</h2>
<ul>
<li>We recommend that you choose a small sample from your data set before attempting to run entire data. and the evaluation metric so it can select the best model. Currently models within “stats” are compared using AIC and BIC. However, models across different types are compared using RMSE. The results of models are shown using RMSE and Normalized RMSE (ratio of RMSE to the standard deviation of actuals).</li>
<li>You must clean the data and not have any missing values. Make sure the target variable is numeric, otherwise, it won’t run. If there is more than one target variable in your data set, just specify only one for now, and if you know the time interval that is in your data, you can specify it. Otherwise it auto-ts will try to infer the time interval on its own.</li>
<li>If you give Auto_Timeseries a different time interval than what the data has, it will automatically resample the data to the given time interval and use the mean of the target for the resampled period.</li>
<li>Notice that except for filename and ts_column input arguments, which are required, all other arguments are optional.</li>
<li>Note that optionally you can give a separator for the data in your file. Default is comma (",").</li>
<li>“time_interval” options are any codes that you can find in this page below.
<a href="https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases">Pandas date-range frequency aliases</a>
</li>
<li>Optionally, you can give seasonal_period as any integer that measures the seasonality in the data. If not given, seasonal_period is assumed automatically as follows: Months = 12, Days = 30, Weeks = 52, Qtr = 4, Year = 1, Hours = 24, Minutes = 60 and Seconds = 60.</li>

CHANGELOG
-------------------------------
version_number = '0.0.24.b2'
-------------------------------
Fixed bug with Prophet Multivariate Prediction where it needed to pass forecast_period in 
addition to X_exogen. Only X_egogen is needed now. Forecast Period is calculated based on 
number of observations in the X_exogen data.

TODO: Make sure all predict functions are consistent (Prophet now has an Optional 
return if things go wrong. Others should do the same.)

-------------------------------
version_number = '0.0.24'
-------------------------------
Added 'auto_arima' capabaility from pmdarima library

-------------------------------
version_number = '0.0.23.b4'
-------------------------------
Changed default argument for 'sep' in fit function to be 'None' (treated as ',' internally).

Fixed bug with predict function in auto_ts 
Dataframe index for X_exogen needed to be set before passing to predict since we were doing the same
while fitting. Without this, it was causing issues with ML models where we are internally 
constructing the 'future dataframe' and if while fiting, the dataframe had datatime index and while
predicting, X_egogen had integer index (index was still in dataframne column in X_egogen), it was
causing issues while adding time series features (could not get time series features from integers).


-------------------------------
version_number = '0.0.23.b3'
-------------------------------
More time series engineered features included in ML models
Example, 'dayofweek', 'quarter', 'month', 'year', 'dayofyear', 'weekofyear', 'weekend', etc.


-------------------------------
version_number = '0.0.23.b2'
-------------------------------
Fixed bug in Prophet rolling window horizon calculation


-------------------------------
version_number = '0.0.23'
-------------------------------
Prophet now includes multivariate modeling capability with rolling window
SARIMAX also includes multivariate modeling capability with rolling window

auto_ts/final.py
####################################################################################
####                     Auto Time Series Final  0.0.14                         ####
####                           Python 3 Version                                 ####
####                      Developed by Ram Seshadri, Kevin Chen                 ####
####                        All Rights Reserved                                 ####
####################################################################################

import pandas as pd
import numpy as np
from datetime import datetime
import warnings
warnings.filterwarnings("ignore")
import seaborn as sns
import matplotlib.pyplot as plt
get_ipython().magic('matplotlib inline')
sns.set(style="white", color_codes=True)
import copy
import ipdb
from collections import defaultdict
import operator
import time
#######################################
# Models
from models import build_arima_model, build_sarimax_model, build_var_model, \
                   build_pyflux_model, build_prophet_model, run_ensemble_model
# Utils
from utils import colorful, load_ts_data, convert_timeseries_dataframe_to_supervised, \
                  time_series_plot, print_static_rmse, print_dynamic_rmse


def Auto_Timeseries(trainfile, ts_column, sep=',', target=None, score_type='rmse',
                    forecast_period=2, timeinterval='', non_seasonal_pdq=None,
                    seasonality=False, seasonal_period=12, seasonal_PDQ=None,
                    conf_int=0.95, model_type="stats", verbose=0):
    """
    ##################################################################################################
    AUTO_TIMESERIES IS A COMPLEX MODEL BUILDING UTILITY FOR TIME SERIES DATA. SINCE IT AUTOMATES MANY
    TASKS INVOLVED IN A COMPLEX ENDEAVOR, IT ASSUMES MANY INTELLIGENT DEFAULTS. BUT YOU CAN CHANGE THEM.
    Auto_Timeseries will rapidly build predictive models based on Statsmodels ARIMA, Seasonal ARIMA
    and Scikit-Learn ML. It will automatically select the BEST model which gives best score specified.
    It will return the best model and a dataframe containing predictions for forecast_period (default=2).
    #####################################################################################################
    INPUT:
    #####################################################################################################
    trainfile: name of the file along with its data path or a dataframe. It accepts both.
    ts_column: name of the datetime column in your dataset (it could be name or number)
    target: name of the column you are trying to predict. Target could also be the only column in your data
    score_type: 'rmse' is the default. You can choose among "mae", "mse" and "rmse".
    forecast_period: default is 2. How many periods out do you want to forecast? It should be an integer
    timeinterval: default is "Month". What is the time period in your data set. Options are: "days",
    model_type: default is "stats". Choice is between "stats", "prophet" and "ml". "All" will build all.
        - stats will build statsmodels based ARIMA< SARIMAX and VAR models
        - ml will build a machine learning model using Random Forests provided explanatory vars are given
        - prophet will build a model using FB Prophet -> this means you must have FB Prophet installed
        - all will build all the above models which may take a long time for large data sets.
    We recommend that you choose a small sample from your data set bedfore attempting to run entire data.
    #####################################################################################################
    and the evaluation metric so it can select the best model. Currently only 2 are supported: RMSE and
    Normalized RMSE (ratio of RMSE to the standard deviation of actuals). Other eval metrics will be soon.
    the target variable you are trying to predict (if there is more than one variable in your data set),
    and the time interval that is in the data. If your data is in a different time interval than given,
    Auto_Timeseries will automatically resample your data to the given time interval and learn to make
    predictions. Notice that except for filename and ts_column which are required, all others are optional.
    Note that optionally you can give a separator for the data in your file. Default is comman (",").
    "timeinterval" options are: 'Days', 'Weeks', 'Months', 'Qtr', 'Year', 'Minutes', 'Hours', 'Seconds'.
    Optionally, you can give seasonal_period as any integer that measures the seasonality in the data.
    If not, seasonal_period is assumed automatically as follows: Months = 12, Days = 30, Weeks = 52,
    Qtr = 4, Year = 1, Hours = 24, Minutes = 60 and Seconds = 60.
    If you want to give your own order, please input it as non_seasonal_pdq and seasonal_PDQ in the input
    as tuples. For example, seasonal_PDQ = (2,1,2) and non_seasonal_pdq = (0,0,3). It will accept only tuples.
    The defaul is None and Auto_Timeseries will automatically search for the best p,d,q (for Non Seasonal)
    and P, D, Q (for Seasonal) orders by searching for all parameters from 0 to 12 for each value of
    p,d,q and 0-3 for each P, Q and 0-1 for D.
    #####################################################################################################
    """

    start_time = time.time()

    ##### Best hyper-parameters in statsmodels chosen using the best aic, bic or whatever. Select here.
    stats_scoring = 'aic'
    seed = 99

    ### If run_prophet is set to True, then only 1 model will be run and that is FB Prophet ##
    lag = copy.deepcopy(forecast_period)-1
    if type(non_seasonal_pdq) == tuple:
        p_max = non_seasonal_pdq[0]
        d_max = non_seasonal_pdq[1]
        q_max = non_seasonal_pdq[2]
    else:
        p_max = 12
        d_max = 1
        q_max = 12
    ################################
    if type(seasonal_PDQ) == tuple:
        seasonal_order = copy.deepcopy(seasonal_PDQ)
    else:
        seasonal_order = (3, 1, 3)

    ########## This is where we start the loading of the data file ######################
    if isinstance(trainfile, str):
        if trainfile != '':
            try:
                ts_df = load_ts_data(trainfile, ts_column, sep, target)
                print('    File loaded successfully. Shape of data set = %s' %(ts_df.shape,))
            except:
                print('File could not be loaded. Check the path or filename and try again')
                return
        elif isinstance(trainfile, pd.DataFrame):
            print('Input is data frame. Performing Time Series Analysis')
            ts_df = load_ts_data(trainfile, ts_column, sep, target)
        else:
            print('File name is an empty string. Please check your input and try again')
            return
    else:
        print('Dataframe given as input. Analyzing Time Series data...')
        ts_df = copy.deepcopy(trainfile)
    df_orig = copy.deepcopy(ts_df)
    if ts_df.shape[1] == 1:
        ### If there is only one column, you assume that to be the target column ####
        target = list(ts_df)[0]
    if not isinstance(ts_column, str):
        ### If ts_column is a number, then it means you need to convert it to a named variable
        ts_column = list(ts_df)[ts_column]
    if isinstance(target,list):
        target = target[0]
        print('    Taking the first column in target list as Target variable = %s' %target)
    else:
        print('    Target variable = %s' %target)
    preds = [x for x in list(ts_df) if x not in [ts_column,target]]

    ##################################################################################################
    ### Turn the time series index into a variable and calculate the difference.
    ### If the difference is not in days, then it is a hourly or minute based time series
    ### If the difference a multiple of days, then test it for weekly, monthly, qtrly, annual etc.
    ##################################################################################################
    if ts_df.index.dtype=='int' or ts_df.index.dtype=='float':
        ### You must convert the ts_df index into a date-time series using the ts_column given ####
        ts_df = ts_df.set_index(ts_column)
    ts_index = ts_df.index

    ################    IF TIME INTERVAL IS NOT GIVEN DO THIS   ########################
    #######   This is where the program tries to tease out the time period in the data set ###########
    ##################################################################################################
    if timeinterval == '':
        ts_index = pd.to_datetime(ts_df.index)
        diff = (ts_index[1] - ts_index[0]).to_pytimedelta()
        diffdays = diff.days
        diffsecs = diff.seconds
        if diffsecs == 0:
            diff_in_hours = 0
            diff_in_days = abs(diffdays)
        else:
            diff_in_hours = abs(diffdays*24*3600 + diffsecs)/3600
        if diff_in_hours == 0 and diff_in_days >= 1:
            print('Time series input in days = %s' % diff_in_days)
            if diff_in_days == 7:
                print('it is a Weekly time series.')
                timeinterval = 'weeks'
            elif diff_in_days == 1:
                print('it is a Daily time series.')
                timeinterval = 'days'
            elif 28 <= diff_in_days < 89:
                print('it is a Monthly time series.')
                timeinterval = 'months'
            elif 89 <= diff_in_days < 178:
                print('it is a Quarterly time series.')
                timeinterval = 'qtr'
            elif 178 <= diff_in_days < 360:
                print('it is a Semi Annual time series.')
                timeinterval = 'qtr'
            elif diff_in_days >= 360:
                print('it is an Annual time series.')
                timeinterval = 'years'
            else:
                print('Time Series time delta is unknown')
                return
        if diff_in_days == 0:
            if diff_in_hours == 0:
                print('Time series input in Minutes or Seconds = %s' % diff_in_hours)
                print('it is a Minute time series.')
                timeinterval = 'minutes'
            elif diff_in_hours >= 1:
                print('it is an Hourly time series.')
                timeinterval = 'hours'
            else:
                print('It is an Unknown Time Series delta')
                return
    else:
        print('Time Interval is given as %s' % timeinterval)

    ################# This is where you test the data and find the time interval #######
    timeinterval = timeinterval.strip().lower()
    if timeinterval in ['months', 'month', 'm']:
        timeinterval = 'months'
        seasonal_period = 12
    elif timeinterval in ['days', 'daily', 'd']:
        timeinterval = 'days'
        seasonal_period = 30
        # Commented out b/c resample only works with DatetimeIndex, not Index
        # ts_df = ts_df.resample('D').sum()
    elif timeinterval in ['weeks', 'weekly', 'w']:
        timeinterval = 'weeks'
        seasonal_period = 52
    elif timeinterval in ['qtr', 'quarter', 'q']:
        timeinterval = 'qtr'
        seasonal_period = 4
    elif timeinterval in ['years', 'year', 'annual', 'y', 'a']:
        timeinterval = 'years'
        seasonal_period = 1
    elif timeinterval in ['hours', 'hourly', 'h']:
        timeinterval = 'hours'
        seasonal_period = 24
    elif timeinterval in ['minutes', 'minute', 'min', 'n']:
        timeinterval = 'minutes'
        seasonal_period = 60
    elif timeinterval in ['seconds', 'second', 'sec', 's']:
        timeinterval = 'seconds'
        seasonal_period = 60
    else:
        timeinterval = 'months'
        seasonal_period = 12

    ########################### This is where we store all models in a nested dictionary ##########
    mldict = lambda: defaultdict(mldict)
    ml_dict = mldict()

    ######### This is when you need to use FB Prophet ###################################
    ### When the time interval given does not match the tested_timeinterval, then use FB.
    #### Also when the number of rows in data set is very large, use FB Prophet, It is fast.
    #########                 FB Prophet              ###################################
    if model_type.lower() == 'prophet' or model_type.lower() == 'all':
        name = 'FB_Prophet'
        print(colorful.BOLD + '\nRunning Facebook Prophet Model...' + colorful.END)
        # try:
        #### If FB prophet needs to run, it needs to be installed. Check it here ###
        model, forecast_df, rmse, norm_rmse = build_prophet_model(
                                    ts_df, ts_column, target, forecast_period,
                                    score_type, verbose, conf_int)
        ml_dict[name]['model'] = model
        ml_dict[name]['forecast'] = forecast_df['yhat'].values
        ##### Make sure that RMSE works, if not set it to np.inf  #########
        if score_type == 'rmse':
            score_val = rmse
        else:
            score_val = norm_rmse
        # except:
        #     print('    FB Prophet may not be installed or Model is not running...')
        #     score_val = np.inf
        ml_dict[name][score_type] = score_val
    elif model_type.lower() == 'stats' or model_type.lower() == 'all':
        ##### First let's try the following models in sequence #########################################
        nsims = 100   ### this is needed only for M-H models in PyFlux
        name = 'PyFlux'
        print(colorful.BOLD + '\nRunning PyFlux Model...' + colorful.END)
        try:
            ml_dict[name]['model'], ml_dict[name]['forecast'], rmse, norm_rmse = \
                build_pyflux_model(ts_df, target, p_max, q_max, d_max, forecast_period,
                                   'MLE', nsims, score_type, verbose)
        except:
            print('    PyFlux model error: predictions not available.')
            score_val = np.inf
            rmse = np.inf
            norm_rmse = np.inf
        ##### Make sure that RMSE works, if not set it to np.inf  #########
        if score_type == 'rmse':
            score_val = rmse
        else:
            score_val = norm_rmse
        ml_dict[name][score_type] = score_val
        ################### Let's build an ARIMA Model and add results #################
        name = 'ARIMA'
        print(colorful.BOLD + '\nRunning Non Seasonal ARIMA Model...' + colorful.END)
        try:
            ml_dict[name]['model'], ml_dict[name]['forecast'], rmse, norm_rmse = build_arima_model(ts_df[target],
                                                    stats_scoring,p_max,d_max,q_max,
                                    forecast_period=forecast_period,method='mle',verbose=verbose)
        except:
            print('    ARIMA model error: predictions not available.')
            score_val = np.inf
        if score_type == 'rmse':
            score_val = rmse
        else:
            score_val = norm_rmse
        ml_dict[name][score_type] = score_val
        ############# Let's build a SARIMAX Model and get results ########################
        name = 'SARIMAX'
        print(colorful.BOLD + '\nRunning Seasonal SARIMAX Model...' + colorful.END)
        # try:
        ml_dict[name]['model'], ml_dict[name]['forecast'], rmse, norm_rmse = build_sarimax_model(ts_df[target], stats_scoring, seasonality,
                                                seasonal_period, p_max, d_max, q_max,
                                                forecast_period,verbose)
        # except:
        #     print('    SARIMAX model error: predictions not available.')
        #     score_val = np.inf
        if score_type == 'rmse':
            score_val = rmse
        else:
            score_val = norm_rmse
        ml_dict[name][score_type] = score_val
        ########### Let's build a VAR Model - but first we have to shift the predictor vars ####
        name = 'VAR'
        if len(preds) == 0:
            print('No VAR model since number of predictors is zero')
            rmse = np.inf
            norm_rmse = np.inf
        else:
            try:
                if df_orig.shape[1] > 1:
                    preds = [x for x in list(df_orig) if x not in [target]]
                    print(colorful.BOLD + '\nRunning VAR Model...' + colorful.END)
                    print('    Shifting %d predictors by 1 to align prior predictor values with current target values...'
                                            %len(preds))
                    ts_df[preds] = ts_df[preds].shift(1)
                    ts_df.dropna(axis=0,inplace=True)
                    ml_dict[name]['model'], ml_dict[name]['forecast'], rmse, norm_rmse = build_var_model(ts_df[[target]+preds],stats_scoring,
                                                forecast_period, p_max, q_max)
                else:
                    print(colorful.BOLD + '\nNo predictors available. Skipping VAR model...' + colorful.END)
                    score_val = np.inf
            except:
                print('    VAR model error: predictions not available.')
                rmse = np.inf
                norm_rmse = np.inf
        ################################################################
        if score_type == 'rmse':
            score_val = rmse
        else:
            score_val = norm_rmse
        ########################################################################
        ml_dict[name][score_type] = score_val
    elif model_type.lower() == 'ml' or model_type.lower() == 'all':
        ########## Let's build a Machine Learning Model now with Time Series Data ################
        name = 'ML'
        if len(preds) == 0:
            print('No ML model since number of predictors is zero')
            rmse = np.inf
            norm_rmse = np.inf
        else:
            # try:
                if df_orig.shape[1] > 1:
                    preds = [x for x in list(ts_df) if x not in [target]]
                    print(colorful.BOLD + '\nRunning Machine Learning Models...' + colorful.END)
                    print('    Shifting %d predictors by lag=%d to align prior predictor with current target...'
                                % (len(preds), lag))
                    # ipdb.set_trace()
                    dfxs, target, preds = convert_timeseries_dataframe_to_supervised(ts_df[preds+[target]],
                                            preds+[target], target, n_in=lag, n_out=0, dropT=False)
                    train = dfxs[:-forecast_period]
                    test = dfxs[-forecast_period:]
                    best = run_ensemble_model(train[preds], train[target], 'TimeSeries',
                                              score_type, verbose)
                    bestmodel = best[0]
                    ml_dict[name]['model'] = bestmodel
                    ### Certain models dont have random state => so dont do this for all since it will error
                    #best.set_params(random_state=0)
                    ml_dict[name]['forecast'] = bestmodel.fit(train[preds],train[target]).predict(test[preds])
                    rmse, norm_rmse = print_dynamic_rmse(test[target].values,
                                                bestmodel.predict(test[preds]),
                                                train[target].values)
                    #### Plotting actual vs predicted for RF Model #################
                    plt.figure(figsize=(5, 5))
                    plt.scatter(train.append(test)[target].values,
                                np.r_[bestmodel.predict(train[preds]), bestmodel.predict(test[preds])])
                    plt.xlabel('Actual')
                    plt.ylabel('Predicted')
                    plt.show()
                else:
                    print(colorful.BOLD + '\nNo predictors available. Skipping Machine Learning model...' + colorful.END)
                    score_val = np.inf
            # except:
            #     print('    For ML model, evaluation score is not available.')
            #     score_val = np.inf
        ################################################################
        if score_type == 'rmse':
            score_val = rmse
        else:
            score_val = norm_rmse
            rmse = np.inf
            norm_rmse = np.inf
        ########################################################################
        ml_dict[name][score_type] = score_val
        ############ Draw a plot of the Time Series data given so you can select p,d,q ######
        time_series_plot(dfxs[target], chart_time=timeinterval)
    else:
        print('No model_type given or it is unknown type. Please look at input and run again')
        return ml_dict
    ######## Selecting the best model based on the lowest rmse score ######
    f1_stats = {}
    for key, val in ml_dict.items():
        f1_stats[key] = ml_dict[key][score_type]
    best_model_name = min(f1_stats.items(), key=operator.itemgetter(1))[0]
    print(colorful.BOLD + '\nBest Model is:' + colorful.END)
    print('    %s' % best_model_name)
    best_model = ml_dict[best_model_name]['model']
    #print('    Best Model Forecasts: %s' %ml_dict[best_model_name]['forecast'])
    print('    Best Model Score: %0.2f' % ml_dict[best_model_name][score_type])
    return ml_dict
##########################################################
#Defining AUTO_TIMESERIES here
##########################################################
if	__name__	== "__main__":
    version_number = '0.0.14'
    print("""Running Auto Timeseries version: %s...Call by using Auto_Timeseries(trainfile, ts_column,
                            sep=',', target=None, score_type='rmse', forecast_period=2,
                            timeinterval='Month', non_seasonal_pdq=None, seasonality=False,
                            seasonal_period=12, seasonal_PDQ=None,
                            verbose=0)
    To get detailed charts of actuals and forecasts, set verbose = 1""" % version_number)
else:
    version_number = '0.0.14'
    print("""Imported Auto_Timeseries version: %s. Call by using Auto_Timeseries(trainfile, ts_column,
                            sep=',', target=None, score_type='rmse', forecast_period=2,
                            timeinterval='Month', non_seasonal_pdq=None, seasonality=False,
                            seasonal_period=12, seasonal_PDQ=None,
                            verbose=0)
    To get detailed charts of actuals and forecasts, set verbose = 1""" % version_number)
